### ISSUE TITLE
Add discontinuities handler to integrator

### ISSUE THREAD

**Issue by drhagen** [drhagen/scipy_ode#11]
As discussed in #2, most ODE solvers are built under some assumptions of the continuity of the RHS and its derivatives. Frequently, there are problems with discontinuities in the RHS that we still want to solve. In an ideal world, the integrator would be able to handle an arbitrary RHS without the user having to list out the discontinuities. However, I know of no method to detect these discontinuities in general. The solution is to solve the system up to the discontinuity and then restart the solver past the discontinuity, but ideally this would be transparent to the user.

Using the step solvers, the user can implement such an integrator on his own. Indeed, I will do it locally if the general consensus is that this does not belong in scipy. However, I think this might be a common enough problem to warrant its inclusion. I have written [a discontinuities handler for Matlab](https://github.com/kroneckerbio/kroneckerbio/blob/master/Source/private/accumulateOdeFwdComp.m) as part of my biological modeling toolbox.

I propose adding a keyword `discontinuities` to `solve_ivp`. It would have a default value of `()`. The integrator would follow this procedure:

1. Initialize the solver with 
    * `t0=t0`
    * `y0=y0`
    * `t_crit=discontinuities[0] - EPS`
2. Step the solver until finished (just a bit before the discontinuity)
3. Initialize a fresh solver with 
    * `t0=discontinuities[0] + EPS` (just a bit after the discontinuity)
    * `y0=y(t_crit)`
    * `t_crit=discontinuities[1] - EPS`
4. Step the solver until finished
5. Continue until there are no more discontinuities or `tF` is reached.

The `OdeSolution` object will store a list of interpolators, one for each each continuous segment. When the solution is queried, it looks up which interpolator segment has the answer and passes the query onto it. If multiple times were queried, it transparently stitches the solution together and returns a simple array.

The advantage of this solution is that we can integrate a discontinuous RHS. The only change to the interface is the `discontinuities` keyword. Everything else happens transparently for the user.

Questions
* Should `discontinuities` be allowed to be any iterable?
    * Advantages; I would probably going to use a for loop anyway. Some people may want to use `range`. If we ever want this feature in `solve_ivp_lazy`, it will have to be an iterable.
    * Disadvantages: We can't preallocate the list of interpolator segments
* If someone evaluates the solution at the tiny gap at a discontinuity, how should that be solved?
    * Return the last state of the previous segment
    * Turn extrapolation on for every interpolator and extrapolate out an EPS with the previous segment
    * Save the state at each discontinuity as a separate field and return it when requested
    * Note that the only difference in these possibilities is elegance. All give the same answer (well, within an EPS).

**Comment by aarchiba** [comment]
Do you have some examples of the kind of discontinuity you're talking about? Is it possible to set t_crit equal to the actual discontinuity and integrate right up to it? Some solvers do require evaluation exactly at the end of the interval, but there should be some that don't.

**Comment by drhagen** [comment]
> Do you have some examples of the kind of discontinuity you're talking about?

Pharmacology: Have a cancer drug infused for an hour each day. Also, each day the subject has random day-to-day variation represented as a saw-tooth function with random heights. Simulate tumor size.

Engineering: Keep the temperature of a system constant, then slowly ramp up the temperature, and hold it at a maximum. Simulate crystal quality.

> Is it possible to set t_crit equal to the actual discontinuity and integrate right up to it?

Not in general. I have made that mistake many times, crashing the integrator. It is easiest to see why this will fail if the RHS is a step function that is right-continuous. The change in the RHS is infinite; the hessian of the RHS is infinite; etc. To integrate this within a given tolerance requires a step size of 0. I am not super familiar with ODE solvers, but I think this is generally how it works.

> Some solvers do require evaluation exactly at the end of the interval, but there should be some that don't.

That is correct. For example, forward Euler doesn't evaluate the last point. But I think a majority of advanced solvers evaluate the last point. If we decide to include this, I don't think there are downsides to stopping an epsilon from a discontinuity for all solvers.

**Comment by aarchiba** [comment]
In a step-it-yourself framework, I'd implement those by using t_crit to make sure the solver stopped at each transition point and then change the RHS return values, essentially making sure that the discontinuity returned one value when evaluated "from the left" and another "from the right". But this is not a convenient way to handle it if you just want the solution. 

The discontinuities you're talking about, if I understand correctly, are step discontinuities in the RHS. The solution remains continuous but not (necessarily) differentiable at that point. Certainly adaptive solvers are going to be forced to small step sizes, but low-order solvers shouldn't be pushed all that hard. When you say "crashing" the solver, what happened? Did it hit a minimum step size/maximum step count? Perhaps this is an application for temporarily accepting a step with higher estimated error and then continuing?

Does it make sense to make a notebook, or a collection of notebooks, of use cases or trial problems, things we would like to make sure are convenient to handle with these tools? I'd be happy to contribute a few illustrating the things I keep asking for. This is not to replace test cases but to guide development.

**Comment by aarchiba** [comment]
Incidentally, in "Solving Ordinary Differential Equations I: Nonstiff problems" by Hairer, Nørsett and Wanner, they talk about what happens when the RHS is discontinuous (not necessarily at a point identifiable ahead of time). They encourage users to detect when this occurs (including using event detection) and stop the solver so that the RHS can be switched, but they also show that simply ignoring the switch is feasible; in particular, using the Dormand-Prince adaptive 5th-order integrator (same one we have) with a tolerance of 1e-5 they show that it manages to navigate past a discontinuity at the cost of some rejected steps:
![ode-discontinuity](https://cloud.githubusercontent.com/assets/136867/18055961/384badf0-6e0a-11e6-814e-5ca3c70fe044.png)

**Comment by drhagen** [comment]
I withdraw this proposal.

I tried to use this today, discovered that its handling of discontinuities was not exactly what I wanted, and implemented something different using the step solvers. I found that I had a right-continuous function, and want my solution to reflect that. I think there are too many possible ways to handle discontinuities to justify baking "one true way" into the convenience functions. If people want this functionality, they can implement a solution tailored to their problem. (It was very easy to do, by the way.) If you want to run over the discontinuities, I guess that's already built in.
